### ISSUE TITLE
compile error

### ISSUE THREAD

**Issue by r3flow** [jodinathan/js_bindings#1]
Great library! Thank you for your work!

Today I tried with booth latest flutter stable and latest beta channel but I got same result with a single line of code in a flutter project (latest android studio, win10): `await window.navigator.mediaDevices.getUserMedia().future;`

```
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/html.dart:201:7: Error: The superclass, 'Element', has no unnamed constructor that takes no arguments.
class HTMLElement // Node -> {web_animations_1: {Animatable, ARIAMixin, GeometryUtils, Region, ParentNode, NonDocumentTypeChildNode, ChildNode, Slottable, InnerHTML}, wai_aria_1_2: {ARIAMixin}, cssom_view_1: {GeometryUtils}, css_regions_1: {Region}, dom: {ParentNode, NonDocumentTypeChildNode, ChildNode, Slottable}, d_o_m_parsing: {InnerHTML}} -> Element
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/html.dart:412:7: Error: The superclass, 'HTMLElement', has no unnamed constructor that takes no arguments.
class HTMLLinkElement // Element -> {html: {GlobalEventHandlers, DocumentAndElementEventHandlers, ElementContentEditable, HTMLOrSVGElement, ElementCSSInlineStyle}, cssom_1: {ElementCSSInlineStyle}} -> HTMLElement
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/html.dart:518:7: Error: The superclass, 'HTMLElement', has no unnamed constructor that takes no arguments.
class HTMLStyleElement // Element -> {html: {GlobalEventHandlers, DocumentAndElementEventHandlers, ElementContentEditable, HTMLOrSVGElement, ElementCSSInlineStyle}, cssom_1: {ElementCSSInlineStyle}} -> HTMLElement
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/html.dart:538:7: Error: The superclass, 'HTMLElement', has no unnamed constructor that takes no arguments.
class HTMLBodyElement // Element -> {html: {GlobalEventHandlers, DocumentAndElementEventHandlers, ElementContentEditable, HTMLOrSVGElement, ElementCSSInlineStyle}, cssom_1: {ElementCSSInlineStyle}} -> HTMLElement
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/html.dart:780:7: Error: The superclass, 'HTMLElement', has no unnamed constructor that takes no arguments.
class HTMLAnchorElement // Element -> {html: {GlobalEventHandlers, DocumentAndElementEventHandlers, ElementContentEditable, HTMLOrSVGElement, ElementCSSInlineStyle}, cssom_1: {ElementCSSInlineStyle}} -> HTMLElement
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/html.dart:2430:7: Error: The superclass, 'HTMLElement', has no unnamed constructor that takes no arguments.
class HTMLAreaElement // Element -> {html: {GlobalEventHandlers, DocumentAndElementEventHandlers, ElementContentEditable, HTMLOrSVGElement, ElementCSSInlineStyle}, cssom_1: {ElementCSSInlineStyle}} -> HTMLElement
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/html.dart:7805:7: Error: The superclass, 'WorkerGlobalScope', has no unnamed constructor that takes no arguments.
class DedicatedWorkerGlobalScope // EventTarget -> {css_font_loading_3: {FontFaceSource, WindowOrWorkerGlobalScope}, html: {WindowOrWorkerGlobalScope}} -> WorkerGlobalScope
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/html.dart:8386:7: Error: The superclass, 'HTMLElement', has no unnamed constructor that takes no arguments.
class HTMLFrameSetElement // Element -> {html: {GlobalEventHandlers, DocumentAndElementEventHandlers, ElementContentEditable, HTMLOrSVGElement, ElementCSSInlineStyle}, cssom_1: {ElementCSSInlineStyle}} -> HTMLElement
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/dom.dart:1233:7: Error: The superclass, 'Node', has no unnamed constructor that takes no arguments.
class Document // EventTarget -> {} -> Node
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/dom.dart:1972:7: Error: The superclass, 'Node', has no unnamed constructor that takes no arguments.
class DocumentType // EventTarget -> {} -> Node
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/dom.dart:2001:7: Error: The superclass, 'Node', has no unnamed constructor that takes no arguments.
class DocumentFragment // EventTarget -> {} -> Node
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/dom.dart:2016:7: Error: The superclass, 'DocumentFragment', has no unnamed constructor that takes no arguments.
class ShadowRoot // Node -> {dom: {NonElementParentNode, ParentNode}} -> DocumentFragment
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/dom.dart:2060:7: Error: The superclass, 'Node', has no unnamed constructor that takes no arguments.
class Element // EventTarget -> {} -> Node
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/dom.dart:2719:7: Error: The superclass, 'Node', has no unnamed constructor that takes no arguments.
class CharacterData // EventTarget -> {} -> Node
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/dom.dart:2785:7: Error: The superclass, 'CharacterData', has no unnamed constructor that takes no arguments.
class Text // Node -> {dom: {NonDocumentTypeChildNode, ChildNode}} -> CharacterData
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/dom.dart:2848:7: Error: The superclass, 'CharacterData', has no unnamed constructor that takes no arguments.
class ProcessingInstruction // Node -> {dom: {NonDocumentTypeChildNode, ChildNode}} -> CharacterData
      ^

...

/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/deprecation_reporting.dart:27:7: Error: JS interop class 'DeprecationReportBody' cannot extend Dart class 'Object with ReportBody'.
Try removing the JS interop annotation or adding it to the parent class.
class DeprecationReportBody // null -> {} -> ReportBody
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/element_timing.dart:23:7: Error: JS interop class 'PerformanceElementTiming' cannot extend Dart class 'Object with PerformanceEntry'.
Try removing the JS interop annotation or adding it to the parent class.
class PerformanceElementTiming // null -> {} -> PerformanceEntry
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/web_nfc.dart:110:7: Error: JS interop class 'NDEFReader' cannot extend Dart class 'Object with EventTarget'.
Try removing the JS interop annotation or adding it to the parent class.
class NDEFReader // null -> {} -> EventTarget
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/web_nfc.dart:156:7: Error: JS interop class 'NDEFReadingEvent' cannot extend Dart class 'Object with Event'.
Try removing the JS interop annotation or adding it to the parent class.
class NDEFReadingEvent // null -> {} -> Event
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/web_nfc.dart:173:7: Error: JS interop class 'NDEFReadingEventInit' cannot extend Dart class 'Object with EventInit'.
Try removing the JS interop annotation or adding it to the parent class.
class NDEFReadingEventInit // null -> {} -> EventInit
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/webrtc_stats.dart:15:6: Error: JS interop classes do not support non-external constructors.
Try annotating with `external`.
enum RTCStatsType {
     ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/webrtc_stats.dart:15:6: Error: This JS interop member must be annotated with `external`. Only factories and static methods can be non-external.
Try annotating the member with `external`.
enum RTCStatsType {
     ^

...

/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/svg11.dart:1317:7: Error: Dart class 'SVGElement with SVGFitToViewBox' cannot extend JS interop class 'SVGElement'.
Try adding the JS interop annotation or removing it from the parent class.
class SVGMarkerElement // Element -> {html: {GlobalEventHandlers, DocumentAndElementEventHandlers, HTMLOrSVGElement, SVGElementInstance}, svg11: {SVGElementInstance}} -> SVGElement
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/svg11.dart:1317:7: Error: JS interop class 'SVGMarkerElement' cannot extend Dart class 'SVGElement with SVGFitToViewBox'.
Try removing the JS interop annotation or adding it to the parent class.
class SVGMarkerElement // Element -> {html: {GlobalEventHandlers, DocumentAndElementEventHandlers, HTMLOrSVGElement, SVGElementInstance}, svg11: {SVGElementInstance}} -> SVGElement
      ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/svg11.dart:1400:7: Error: Dart class 'SVGElement with SVGURIReference' cannot extend JS interop class 'SVGElement'.
Try adding the JS interop annotation or removing it from the parent class.
class SVGGradientElement // Element -> {html: {GlobalEventHandlers, DocumentAndElementEventHandlers, HTMLOrSVGElement, SVGElementInstance}, svg11: {SVGElementInstance}} -> SVGElement
      ^

...

/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/webxr_ar_module_1.dart:23:6: Error: JS interop classes do not support non-external constructors.
Try annotating with `external`.
enum XREnvironmentBlendMode {
     ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/webxr_ar_module_1.dart:23:6: Error: This JS interop member must be annotated with `external`. Only factories and static methods can be non-external.
Try annotating the member with `external`.
enum XREnvironmentBlendMode {
     ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/webxr_ar_module_1.dart:31:6: Error: JS interop classes do not support non-external constructors.
Try annotating with `external`.
enum XRInteractionMode {
     ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/webxr_ar_module_1.dart:31:6: Error: This JS interop member must be annotated with `external`. Only factories and static methods can be non-external.
Try annotating the member with `external`.
enum XRInteractionMode {
     ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/content_index.dart:16:6: Error: JS interop classes do not support non-external constructors.
Try annotating with `external`.
enum ContentCategory {
     ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/content_index.dart:16:6: Error: This JS interop member must be annotated with `external`. Only factories and static methods can be non-external.
Try annotating the member with `external`.
enum ContentCategory {
     ^
/D:/.../pub_cache/hosted/pub.dartlang.org/js_bindings-0.0.3/lib/bindings/resource_timing_2.dart:36:7: Error: JS interop class 'PerformanceResourceTiming' cannot extend Dart class 'Object with PerformanceEntry'.
Try removing the JS interop annotation or adding it to the parent class.
class PerformanceResourceTiming // null -> {} -> PerformanceEntry
      ^
Failed to compile application.
```

**Comment by jodinathan** [comment]
hi, sorry for the late answer.  

The package needs to be updated with the info from the dart team regarding JS interop:  
https://github.com/dart-lang/sdk/issues/46359 and https://github.com/dart-lang/sdk/issues/35084#issuecomment-861848963. 

Basically we need to change the generator to generate bindings to js_utils instead of package:js.  
With that it may be possible to use it with web components :)

Unfortunately I don't have a date to give.  

Of course you are free to make a PR =]

**Comment by anujiotech** [comment]
Not able to generate bindings using tools/scripts.
error related to "atob()" & "self." names.

alternatively if you could let us know last working version of flutter/dart with you plugin

**Comment by ditman** [comment]
Hey, I just found out about this library and I think it's great! I've written by hand parts of this, and would love to use this instead! 😛 

@jodinathan I'm sure you're already aware, but there's some new features coming to Dart (2.16) and `package:js` (0.6.4) that might fix the issue reported above without a full rewrite. In any case, I'm going to plop the links here:

* https://pub.dev/packages/js/versions/0.6.4#interop-with-native-types-using-staticinterop
* https://github.com/dart-lang/sdk/blob/main/sdk/lib/html/doc/WORKAROUNDS.md

The `@staticInterop` annotation looks very handy:

> Previously, you could not use `@JS()` or `@anonymous` types to interface with native types that were reserved within `dart:html` e.g. `Window`.
>
> Using `@staticInterop` will now let you do so. ...

Thanks for this library, and keep it up! It has just shot to the top of my flutter-favorites! 😄

**Comment by jodinathan** [comment]
@ditman yes!  

I am waiting for sdk 2.16 and js `@staticInterop` to test and publish a new version.  
Hopefully with a working example of web components usage

**Comment by jodinathan** [comment]
I've fixed the builders and released a `0.0.4-dev` with the example folder working.  

@ditman, the `@staticInterop` with static method extensions didn't work. I had to to use [`js_util` methods](https://github.com/jodinathan/js_bindings/blob/278f1336771e7ef9b8dc8fd5f3d512030ab7bef3/lib/bindings/dom.dart#L84).

Now I will work on a web components example. Not sure if it is possible yet.

**Comment by jodinathan** [comment]
While I am waiting for some answer regarding the web components usage, I've changed the builders to use more of the `js_util` stuff.  

So now there is no need to call `promiseToFuture` nor `allowInterop` anymore. I've changed the builder to already add these methods calls when necessary. [CHANGELOG](https://github.com/jodinathan/js_bindings/blob/main/CHANGELOG.md) 

[The example](https://github.com/jodinathan/js_bindings/blob/main/example/main.dart) is now more complete with these changes. I have published a new dev version to the pub. 

Would be nice to have some feedback about these changes.  

Also, PLEASE, thumbs up [this issue](https://github.com/w3c/webref/issues/277) in the w3c repo. Once it is done the package will need less manual checks and no need of merging IDL files. Will be much easier to maintain and add stuff.

**Comment by ditman** [comment]
> Also, PLEASE, thumbs up this issue in the w3c repo. Once it is done the package will need less manual checks and no need of merging IDL files. Will be much easier to maintain and add stuff.

@jodinathan it seems they just closed-fixed the issue? https://github.com/w3c/webref/issues/277#issuecomment-1025972132

**Comment by jodinathan** [comment]
it seems so =]

Today I have to make some magic just to get info from the spec right. 

Will check if stuff are ok in the next few days

**Comment by jodinathan** [comment]
published 0.0.4 with several new stuff =] 

create issues if you find anything
